ec/ecresource/deploymentresource/topology_dedicated_masters.go

```diff
@@ -0,0 +1,365 @@
+	"github.com/elastic/cloud-sdk-go/pkg/models"
+	es "github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/elasticsearch/v2"
+	"github.com/elastic/terraform-provider-ec/ec/internal/util"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+var objectAsOptions = basetypes.ObjectAsOptions{UnhandledNullAsEmpty: false, UnhandledUnknownAsEmpty: false}
+
+func UpdateDedicatedMasterTier(
+	ctx context.Context,
+	config tfsdk.Config,
+	plan tfsdk.Plan,
+	privateState PrivateState,
+	resp *resource.ModifyPlanResponse,
+	loadTemplate func() (*models.DeploymentTemplateInfoV2, error),
+) {
+	var esConfig es.ElasticsearchTF
+	resp.Diagnostics.Append(config.GetAttribute(ctx, path.Root("elasticsearch"), &esConfig)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !esConfig.MasterTier.IsNull() && !esConfig.MasterTier.IsUnknown() {
+		// Master tier is explicitly configured -> No changes will be made
+		tflog.Debug(ctx, "Skip UpdateDedicatedMasterTier: Master tier has been explicitly configured")
+		return
+	}
+
+	var planElasticsearch es.ElasticsearchTF
+	resp.Diagnostics.Append(plan.GetAttribute(ctx, path.Root("elasticsearch"), &planElasticsearch)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	template, err := loadTemplate()
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get deployment-template", "Error: "+err.Error())
+		return
+	}
+
+	dedicatedMastersThreshold := getDedicatedMastersThreshold(*template)
+	if dedicatedMastersThreshold == 0 {
+		// No automatic dedicated masters management
+		return
+	}
+
+	nodesInCluster := countNodesInCluster(ctx, planElasticsearch, *template)
+
+	if nodesInCluster < dedicatedMastersThreshold {
+		// Disable master tier
+		if planElasticsearch.MasterTier.IsUnknown() || planElasticsearch.MasterTier.IsNull() {
+			resp.Plan.SetAttribute(ctx,
+				path.Root("elasticsearch").AtName("master"),
+				types.ObjectNull(es.ElasticsearchTopologyAttrs()),
+			)
+		} else {
+			resp.Plan.SetAttribute(ctx,
+				path.Root("elasticsearch").AtName("master").AtName("size"),
+				"0g",
+			)
+		}
+	} else {
+		var migrateToLatestHw bool
+		plan.GetAttribute(ctx, path.Root("migrate_to_latest_hardware"), &migrateToLatestHw)
+
+		// Skip update if the master tier is already enabled
+		// If migrateToLatestHw is true, update the tier to values from latest IC
+		if masterTierIsEnabled(ctx, planElasticsearch, *template) && !migrateToLatestHw {
+			return
+		}
+
+		// Enable master tier
+
+		instanceConfigurations, diags := ReadPrivateStateInstanceConfigurations(ctx, privateState)
```

Comment:
Can we use IC from private state if migration to the latest hardware is available or IC or IC version has changed in plan?

Reply:
Maybe we can if IC version changed (if the IC list in deployment contains all the versions of the used IC) but still does it make sense to read ICs from `loadTemplate` instead (e.g. if template changed)? Also, what shall we do if IC changed?

Reply:
The private state IC is already not used when migrating, but it isn't straightforward to see in the code (Maybe I can make it more readable):

- The current IC can only be read and used if the tier has an ic-id set
- When migrate-hw is set, the tier is actually set to unknown so we don't know the ic-id (This is due to the `UseTopologyStateForUnknown` modifier)
- since we don't know the ic-id, the code falls back to the template

So it works because `getInstanceConfiguration` below always returns `nil` if `migrateToLatestHw` is set.

Reply:
Actually this could be made more straightforward by directly using the template if `migrateToLatestHw==true`

Reply:
I refactored the instance-config selection into a separate function and added an if-case for migration.

Reply:
> When migrate-hw is set, the tier is actually set to unknown so we don't know the ic-id (This is due to the UseTopologyStateForUnknown modifier)

`migrate-hw` just reflects user's desire to move, the actual possibility to move is defined by this flag and the availability of a newer IC version. IIUC `migrate-hw` can be kept in TF config for some time.

From other side, we can document this behavior. Or check that the actual migration is going to take place.

Reply:
>  IIUC migrate-hw can be kept in TF config for some time.

Yeah that works fine with the code. The flag basically means 'always use latest IC from template'.
But it doesn't cause any problems in the code. As long as the template has the same IC there will be no changes, as soon as a new IC version is released the next plan will contain changes.

Reply:
@gigerdo , thanks a lot for the explanations.